### PR TITLE
Removes constraints from parseField method default implementation.

### PR DESCRIPTION
**Pull request by ghost**
Hello. Great library! This PR has a couple of changes, one of them simply removes a constraint that is technically not needed by the affected code. While the other refactors some code in order to remove another constraint. Wondering if you would like to merge one of these or both of these changes. Thanks.

The goal is to make it more convenient to reuse the parseField method by avoiding superfluous constraints on the default implementation.

When readField was added, the Read constraint on parseField is not required by the default implementation any more. I found myself needing to make dummy Read instances in order to reuse the default implementation.

The Typeable constraint is only used to derive the metavar within parseField. So I extracted metavar out as another type class method. A nice side effect of this is that now it's possible to customize the metavar and still reuse the default parseFields implementation. Using Proxy seemed like the right way to implementing this method, but I'm pretty new to type-level programming and generics, so I might be wrong.

### CODE REVIEW

**Review by Gabriella439 — APPROVED** [review]
This is really nice!  I'm just waiting for CI to go green before merging

**Comment by Gabriella439** [comment]
Judging from CI it looks like this might require an explicit `metavar` definition for `Any` and `All` since they don't have a `Typeable` instance for GHC 7.8.4 and older

**Comment by ghost** [comment]
Ah, I forgot to try older versions of GHC. I'll fix it.

**Comment by Gabriella439** [comment]
Awesome!  Thank you :)

I can release this to Hackage if you don't have any other pending changes in mind

**Comment by ghost** [comment]
I don't have anything else in mind. Thanks :)

**Comment by Gabriella439** [comment]
You're welcome!

**Pull request merged**
